packages/dev/core/src/XR/features/WebXRDepthSensing.ts

```diff
@@ -27,10 +36,249 @@ export interface IWebXRDepthSensingOptions {
+            uniformBuffer.setTexture("ds_depthSampler", depthTexture);
+            uniformBuffer.updateMatrix("ds_uvTransform", uvTransform);
+        }
+    }
+
+    public override getClassName() {
+        return "DepthSensingMaterialPlugin";
+    }
+
+    public override getCustomCode(shaderType: string): Nullable<{ [pointName: string]: string }> {
+        return shaderType === "vertex"
+            ? {
+                  CUSTOM_VERTEX_MAIN_BEGIN: `
+                #ifdef MULTIVIEW
+                    ds_viewIndexMultiview = float(gl_ViewID_OVR);
+                #endif
+                `,
+                  CUSTOM_VERTEX_DEFINITIONS: `
+                    varying float ds_viewIndexMultiview;
+                `,
+              }
+            : {
+                  CUSTOM_FRAGMENT_DEFINITIONS: `
+                    #ifdef DEPTH_SENSING
+                        #ifdef DEPTH_SENSING_TEXTURE_ARRAY
+                            uniform highp sampler2DArray ds_depthSampler;
+                        #else
+                            uniform sampler2D ds_depthSampler;
+                        #endif
+                        #ifdef MULTIVIEW
+                            varying float ds_viewIndexMultiview;
+                        #endif
+                    #endif
+                  `,
+                  CUSTOM_FRAGMENT_MAIN_BEGIN: `
+#ifdef DEPTH_SENSING
+    #ifdef MULTIVIEW
+        float ds_viewIndexSet = ds_viewIndexMultiview;
+        vec2 ds_compensation = vec2(0.0, 0.0);
+    #else
+        float ds_viewIndexSet = ds_viewIndex;
+        vec2 ds_compensation = vec2(ds_viewIndexSet, 0.0);
+    #endif
+    vec2 ds_baseUv = gl_FragCoord.xy * ds_invScreenSize;
+    #ifdef DEPTH_SENSING_TEXTURE_ARRAY
+        vec2 ds_uv = ds_baseUv - ds_compensation;
+        vec3 ds_depthUv = vec3((ds_uvTransform * vec4(ds_uv, 0.0, 1.0)).xy, ds_viewIndexSet);
+    #else
+        vec2 ds_depthUv = (ds_uvTransform * vec4(ds_baseUv.x, 1.0 - ds_baseUv.y, 0.0, 1.0)).xy;
+    #endif
+    #ifdef DEPTH_SENSING_TEXTURE_AL
+        // from alpha-luminance - taken from the explainer
+        vec2 ds_alphaLuminance = texture(ds_depthSampler, ds_depthUv).ra;
+        float ds_cameraDepth = dot(ds_alphaLuminance, vec2(255.0, 256.0 * 255.0));
+    #else
+        float ds_cameraDepth = texture(ds_depthSampler, ds_depthUv).r;
+    #endif
+
+    ds_cameraDepth = ds_cameraDepth * ds_rawValueToMeters;
+
+    float ds_assetDepth = gl_FragCoord.z;
+    #ifdef DEPTH_SENSING_DISCARD
+    if(ds_cameraDepth < ds_assetDepth) {
+        discard;
+    }
+    #endif
+#endif  
+                  `,
+                  CUSTOM_FRAGMENT_BEFORE_FRAGCOLOR: `
+#ifdef DEPTH_SENSING
+    #ifndef DEPTH_SENSING_DISCARD
+        const float ds_depthTolerancePerM = 0.005;
+        float ds_occlusion = clamp(1.0 - 0.5 * (ds_cameraDepth - ds_assetDepth) / (ds_depthTolerancePerM * ds_assetDepth) +
+            0.5, 0.0, 1.0);
+        ${this._varColorName} *= (1.0 - ds_occlusion);

```

Comment:
Just to be sure, it's the whole color which is multiplied and not only alpha?

Reply:
yes. this was taken from an example by oculus themselves. I like the idea - it makes the object "fade" behind under certain conditions.